File: py/haltrule/messages.py
# Words for people. Nothing here decides anything: a verdict's `message` and a
# refusal's text are not part of conformance, so no fixture can tell one
# wording from another - and the mutation run leaves this file alone for that
# reason. Code that only words a message lives here; code that decides lives
# in the part.

from __future__ import annotations

from typing import Optional, Union


def show(used: int, limit: Optional[int]) -> str:
    """An amount against its cap."""
    return f"{used} of {'no cap' if limit is None else limit}"


def describe_number(value: Union[int, float]) -> str:
    """A number as a message may hold it. Spelling out an integer past
    Python's int-to-str digit limit raises, and a verdict must not, so a long
    one is described by its size."""
    if isinstance(value, float) or value.bit_length() <= 64:
        return f"{value!r}"
    return f"a {value.bit_length()}-bit integer"

File: rust/haltrule/src/messages.rs
//! Words for people. Nothing here decides anything: a verdict's `message` and
//! a refusal's text are not part of conformance, so no fixture can tell one
//! wording from another. Code that only words a message lives here; code that
//! decides lives in the part.

use alloc::format;
use alloc::string::String;

use crate::value::Value;

pub(crate) fn show_amount(used: i64, limit: Option<i64>) -> String {
    match limit {
        None => format!("{} of no cap", used),
        Some(limit) => format!("{} of {}", used, limit),
    }
}

pub(crate) fn describe_value(value: &Value) -> &'static str {
    // The other ports end this with a catch-all "a value". There is no arm
    // for it here: the enum is closed, so these are all the values there are.
    match value {
        Value::Null => "nothing",
        Value::Bool(_) => "a boolean",
        Value::Int(_) | Value::Float(_) => "a number",
        Value::Text(_) => "a string",
        Value::List(_) => "a list",
        Value::Map(_) => "a map",
    }
}
